tests/callbacks/test_device_stats_monitor.py

```diff
@@ -0,0 +1,129 @@
+from tests.helpers.runif import RunIf
+
+
+@RunIf(min_torch="1.8")
+@RunIf(min_gpus=1)
+def test_device_stats_gpu_from_torch(tmpdir):
+    """Test GPU stats are logged using a logger with Pytorch >= 1.8.0."""
+    model = BoringModel()
+    device_stats = DeviceStatsMonitor()
+    logger = CSVLogger(tmpdir)
+    log_every_n_steps = 2
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=2,
+        limit_train_batches=7,
+        log_every_n_steps=log_every_n_steps,
+        gpus=1,
+        callbacks=[device_stats],
+        logger=logger,
+    )
+
+    trainer.fit(model)
+    assert trainer.state.finished, f"Training failed with {trainer.state}"
+
+    path_csv = os.path.join(logger.log_dir, ExperimentWriter.NAME_METRICS_FILE)
+    met_data = np.genfromtxt(path_csv, delimiter=",", names=True, deletechars="", replace_space=" ")
+
+    fields = ["allocated_bytes.all.freed", "inactive_split.all.peak", "reserved_bytes.large_pool.peak"]
+
+    for f in fields:
+        assert any(f in h for h in met_data.dtype.names)
+
+
+@RunIf(max_torch="1.7")
+@RunIf(min_gpus=1)
+def test_device_stats_gpu_from_nvidia(tmpdir):
+    """Test GPU stats are logged using a logger with Pytorch < 1.8.0."""
+    model = BoringModel()
+    device_stats = DeviceStatsMonitor()
+    logger = CSVLogger(tmpdir)
+    log_every_n_steps = 2
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=2,
+        limit_train_batches=7,
+        log_every_n_steps=log_every_n_steps,
+        gpus=1,
+        callbacks=[device_stats],
+        logger=logger,
+    )
+
+    trainer.fit(model)
+    assert trainer.state.finished, f"Training failed with {trainer.state}"
+
+    path_csv = os.path.join(logger.log_dir, ExperimentWriter.NAME_METRICS_FILE)
+    met_data = np.genfromtxt(path_csv, delimiter=",", names=True, deletechars="", replace_space=" ")
+
+    fields = ["utilization.gpu", "memory.used", "memory.free", "utilization.memory"]
+
+    for f in fields:
+        assert any(f in h for h in met_data.dtype.names)
+
+
+@RunIf(tpu=True)
+def test_device_stats_monitor_tpu(tmpdir):
+    """Test TPU stats are logged using a logger."""
+
+    model = BoringModel()
+    device_stats = DeviceStatsMonitor()
+    logger = CSVLogger(tmpdir)
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=2,
+        limit_train_batches=5,
+        tpu_cores=8,
+        callbacks=[device_stats],
+        logger=logger,
+    )
+
+    trainer.fit(model)
+    assert trainer.state.finished, f"Training failed with {trainer.state}"
+
+    path_csv = os.path.join(logger.log_dir, ExperimentWriter.NAME_METRICS_FILE)
+    met_data = np.genfromtxt(path_csv, delimiter=",", names=True, deletechars="", replace_space=" ")
+
+    fields = ["avg. free memory (MB)", "avg. peak memory (MB)"]
+
+    for f in fields:
+        assert any(f in h for h in met_data.dtype.names)
+
+
+@RunIf(tpu=True)
+def test_device_stats_monitor_tpu_no_logger(tmpdir):
```

Comment:
this test is not TPU specific. we should verify this behavior on CPU so its one less dependency on special tests